### ISSUE TITLE
Is there an error on training Gaussian Naive Bayes for product categorization?

### ISSUE THREAD

**Issue by cioroianudenis** [RubixML/ML#82]
First of all, amazing library, really appreciate what you are doing for the php ecosystem. Keep it up!

I was wondering, I have the following model: 

```
        $samples = [];
        $labels = [];

        $categories = Category::has('variants', '>', 50)->with('variants')->get()->map(function($category) {
            $category->setRelation('variants', $category->variants->take(50));
            return $category;
        });

        foreach ($categories as $label) {
            $tmp_samples = [];
            foreach ($label->variants as $variant) {
                $tmp_samples[] = $variant->title;
            }
            $samples[] = $tmp_samples;
            $labels[] = $label->slug;
        }
        $dataset = new Labeled($samples, $labels);
        $estimator = new PersistentModel(
            new Pipeline([
                new TextNormalizer(),
                new StopWordFilter(StopWords::$words),
                new WordCountVectorizer(10000, 1, new NGram(1, 2, new WordStemmer('romanian'))),
                new TfIdfTransformer(),
                new ZScaleStandardizer(),
            ], new GaussianNB()),
            new Filesystem($this->path.'/categorization.model', true)
        );

        $estimator->train($dataset);
        $estimator->save();
```
which trains on 20000+ product titles (`$samples`) that correspond to their assigned categories (`$lables`). 
After I train it, given a new product title that may or may not have all the words in that original training data vocabulary, I want to make a prediction on which category it best fits.
The code is the following:
```
        $persister = new Filesystem($this->path.'/categorization.model');
        $estimator = $persister->load();

        $prod = ProductVariant::select('title')->find($id);
        $samples[] = [$prod->title];
        $dataset = new Unlabeled($samples);
        dump($estimator->predict($dataset));
```
but I always get the following error after I run `$estimator->predict($dataset)`
`vendor\rubix\ml\src\Transformers\WordCountVectorizer.php:190`
```
ErrorException
Undefined offset: 1
```
corresponding to the following code:
` $tokens = $this->tokenizer->tokenize($sample[$column]);`

Is there a problem with my approach to solving this?
I think there might be something wrong with the way that I have configured the pipeline but I can't seem to figure it out. Any help would be much appreciated. Thank you.

**Comment by andrewdalpino** [comment]
Hi @cioroianudenis thanks for the question!

It appears that your code is correct but the dataset used to train the estimator may have different features, different ordering of features, and/or a different number of feature columns. I suspect this because the fitted [Word Count Vectorizer](https://docs.rubixml.com/en/latest/transformers/word-count-vectorizer.html) is looking for a feature column with offset 1, however, it looks like your dataset for inference only has a single text blob column (only offset 0 exists).

Can you confirm that the training and test/inference sets both have the same dimensionality and features are ordered the same? The example below tests if both sets have the same dimensionality.

```php
$training->numColumns() === $testing->numColumns();
```

Also, I noticed you are using a stemmer, we offer a [Word Stemmer](https://github.com/RubixML/Extras/blob/master/src/Other/Tokenizers/WordStemmer.php) tokenizer as part of the [Extras](https://github.com/RubixML/Extras) package if you wanted to use that along with the [NGram](https://docs.rubixml.com/en/latest/other/tokenizers/n-gram.html) tokenizer.

**Example**

```php
use Rubix\ML\Transformers\WordCountVectorizer;
use Rubix\ML\Other\Tokenizers\NGram;
use Rubix\ML\Other\Tokenizers\WordStemmer;

$transformer = new WordCountVectorizer(10000, 3, new NGram(1, 2, new WordStemmer('english')));
```

It's fine if a word is found during inference that is not part of the vocabulary of Word Count Vectorizer that was fitted during training. It won't contribute any information to the sample, however. In the extreme case that no words are found in the vocabulary it will produce a zero vector.

Also note that [Word Count Vectorizer](https://docs.rubixml.com/en/latest/transformers/word-count-vectorizer.html) received an additional parameter (max document frequency) recently and will be included in the 0.1.0-rc5 release. this parameter allows you to limit the amount of common words that enter the vocabulary, similarly to how min document frequency allows you to limit words that are rare. This will be a breaking change for your code due to the new constructor. Just a heads up!

**Comment by cioroianudenis** [comment]
@andrewdalpino thank you for the quick response!
I have edited my above answer a little bit showing more of the training code that I use to train the model.

The only difference I can think of between the training data which uses the `Labeled` class and the test data which uses the `Unlabeled` class is that the training contains multiple product titles(in the number of thousands) associated with the corresponding category that I manually set, while the test data only contains the title of one product with no category associated, both the training data and test data get pre-processed by the same transformer classes. The product titles can also differ in length size.

I think the problem might be the testing data that should be processed differently in order to obtain the same amount of columns/rows needed in the predicting phase?
I don't exactly know where to put this code `$training->numColumns() === $testing->numColumns();` in order to test that.

Also I noticed that a Stemmer is included in the package but I didn't get the chance to try it out yet, the one that I have is from an old code that I wrote, but I will most definitely switch to the one included in the package.

**Comment by andrewdalpino** [comment]
Word Count Vectorizer will create a vocabulary for *each* categorical feature column of the training set. If you have multiple titles for each training sample, you can either break it into multiple samples with the same label or concatenate the titles into just one feature column. This way you only build one vocabulary (only one is needed) and your input dimensionalities match.

In this example, `$training` is the training dataset and `$testing` is the unlabeled unknown/testing set. You would instantiate both datasets and use the following to check if they have the same number of columns.

```php
$training->numColumns() === $testing->numColumns();
```

**Comment by cioroianudenis** [comment]
@andrewdalpino dude, you simply rock! It works like a charm! Exactly that was the problem! Thank you so much!

**Comment by andrewdalpino** [comment]
Likewise @cioroianudenis! Glad I can help

I'm wondering if we should add some 'dimensionality' checks somewhere in the pipeline

That would probably have been helpful to you in this case no?

**Comment by cioroianudenis** [comment]
@andrewdalpino yeah, I can see that would help with this sort of problems.
